Move to ANTLR 4 with Java 11 language features and localization of more syntax errors.

sampottinger opened the pull request:
Introduces ANTLR4 and Java 8 language feature support within IDE while also adding additional hooks for localization of syntax error messages, addressing https://github.com/processing/processing/issues/3054 and https://github.com/processing/processing/issues/3055.

The PR is broadly a continuation of https://github.com/processing/processing/issues/3055, bringing it up to speed with the latest Processing master plus the changes introduced at https://github.com/processing/processing/pull/5753. **Requires https://github.com/processing/processing/pull/5753 as pre-requisite.** This introduces a number of edits beyond https://github.com/processing/processing/issues/3055 beyond compatibility with current Processing master and https://github.com/processing/processing/pull/5753 including:

 - Update to the grammar itself
 - Change ANTLR listeners to emit `TextTransform.Edit` to unify JDT-based `PreprocessingService` and `JavaBuild`, removing code with duplicate purpose.
 - Introduction of syntax error rewriting with support for localization.
 - Addition of complete localized strings set for English and Spanish.
 - Addition of partial localized strings set for other languages.
 - Refactor of ANTLR-related code for testability and readability
 - Expansion of tests including full parse tests for new Java features (type inference, lambdas)

JakubValtar committed:
ANTLR 4.5 initial commit

based on work of Florian Jenett

- included ANTLR 4.5, building PDE and running sketches works
- error reporting is not yet implemented
- parses size() and adds sketchXxxx methods

JakubValtar committed:
Preproc test cleanup

- moved tests to java mode folder
- fixed expected whitespace in tests
- added sketchXxxx methods to expected test outputs
- error reporting tests are disabled, other tests are gree

sampottinger committed:
Merge branch 'feature-antlr-4.5' of https://github.com/JakubValtar/processing into JakubValtar-feature-antlr-4.5

sampottinger committed:
Fix merge conflicts on sampottinger#15.

sampottinger committed:
Cleanup for java build related to new ANTLR

sampottinger committed:
Working but noisy ANTLR 4 with Java 11.

sampottinger committed:
Working non-syntax errors reporting.

sampottinger committed:
Passable syntax error checking.

sampottinger committed:
Default to STATIC if mode cannot be determined.

sampottinger committed:
Compile with 1.8 language features.

sampottinger commented:
Ohhh man. 🎉 I feel alive. This is working though some of the error messages may require a little clean up. See: https://gist.github.com/sampottinger/45f63c1360a6622e9cb59bea978d18eb.

<img width="812" alt="java8_screenshot" src="https://user-images.githubusercontent.com/110391/55039680-60fe6a00-4fe2-11e9-8a5e-8ef484b88931.png">

sampottinger committed:
Fix treatment of size in antlr4 with settings method.

sampottinger committed:
Fix javadoc to use antlr 4.5 complete.

sampottinger committed:
Remove dead code deprecated by move to ANTLR 4.

sampottinger committed:
Moved tests for compiler as part of ANTLR 4 refactor.

sampottinger committed:
Restored parser, autoformat, and supporting test files.

sampottinger committed:
Finished restoring unit tests for preproc.

sampottinger committed:
Added generics and labmda test.

sampottinger committed:
Introduce updated Java g4 file.

sampottinger committed:
Working primitive function calls  again.

sampottinger committed:
Finished restoring tests under updated grammer.

sampottinger committed:
Cleanup related to updated java grammer.

sampottinger committed:
Merge pull request #20 from sampottinger/move_java_8_antlr

Migrate the Java 8 antlr file.

sampottinger committed:
Fix color type.

sampottinger committed:
Added some error cleanup to support kinder messages.

sampottinger committed:
ANTLR for preprocessing and some related cleanup.

Some style clean up, added error msg handling, and a switch to using ANTLR for all preprocessing within the PreprocessingService.

sampottinger committed:
Allow multi-line grammatical errors.

sampottinger committed:
Minor fixes in ANTLR-based preprocessing after testing.

sampottinger committed:
Fix detecting identifier starting with a number.

sampottinger committed:
Fix 371691f0d4abc3a81552a46e3ca6ed34b145c483.

sampottinger committed:
Fix issue in error formulation in ANTLR for ErrorTable.

sampottinger committed:
Minor edits plus documentation for new preproc contributions.

sampottinger committed:
Minor improvements to issues re extraneous and missing token.

sampottinger committed:
Added line offset utilities for error reporting.

sampottinger committed:
Use original line numbers for known missing.

sampottinger committed:
Fix prior token finding.

sampottinger committed:
Second batch of documentation for sampottinger#15

sampottinger committed:
Start refactor for the ANTLR visitor for code quality and test.

sampottinger committed:
Restored tests after refactor of PdeParseTreeListener.java.

sampottinger committed:
Finished documentation related to ANTLR-required edits.

sampottinger committed:
Finished some documentation in pdex related to new ANTLR.

sampottinger committed:
Finish non-test quality edits related to new ANTLR migration.

sampottinger committed:
Merge pull request #24 from sampottinger/refactor_antlr_visitor

Refactor for the ANTLR visitor for code quality and test.

sampottinger committed:
Added localization support in preproc error mapping.

sampottinger committed:
Fix java line offset for ANTLR.

sampottinger committed:
Clean up reporting of SketchException in JavaBuild / Editor.

sampottinger committed:
Added specific localization functionality for some syntax errors.

Added specific funcationality for localization for syntax error and have included text for English and Spanish.

sampottinger committed:
Added limited syntax error localization support for other lang.

sampottinger committed:
Added missing localization in PDE_el

sampottinger committed:
Added additional testing related to new ANTLR implementation.

sampottinger commented:
Dropping the `[WIP]` part of this PR title. This is ready for review.

sampottinger committed:
Minor fix in TabLineFactoryTest.

sampottinger commented:
Hey there! One more thing... I have included English and Spanish localizations (espro que mi traducción para los mensajes de errors es suficiente) of some error messages along with limited support for other languages using existing strings or (properly attributed) wikipedia translation of short phrases as comfortable. I have also updated #18 to include wikipedia attribution for other languages.

sampottinger committed:
Small typo fix in es editor.status.bad.parameter

monkstone commented:
Hi I just wanted to share with you following ParserTests failures on my linux build 
Apache Ant(TM) version 1.10.5 compiled on September 9 2018
openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment (build 11.0.3+4)
OpenJDK 64-Bit Server VM (build 11.0.3+4, mixed mode):-
```
test:
    [junit] Testsuite: processing.mode.java.AutoFormatTests
    [junit] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.067 sec
    [junit] 
    [junit] Testsuite: processing.mode.java.ParserTests
    [junit] Tests run: 39, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 53.808 sec
    [junit] 
    [junit] Testcase: bug820(processing.mode.java.ParserTests):	FAILED
    [junit] java.lang.IllegalStateException: No match found
    [junit] junit.framework.AssertionFailedError: java.lang.IllegalStateException: No match found
    [junit] 	at processing.mode.java.ParserTests.expectCompilerException(ParserTests.java:81)
    [junit] 	at processing.mode.java.ParserTests.bug820(ParserTests.java:203)
    [junit] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [junit] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 
    [junit] 
    [junit] Testcase: bug1145(processing.mode.java.ParserTests):	FAILED
    [junit] java.lang.IllegalStateException: No match found
    [junit] junit.framework.AssertionFailedError: java.lang.IllegalStateException: No match found
    [junit] 	at processing.mode.java.ParserTests.expectCompilerException(ParserTests.java:81)
    [junit] 	at processing.mode.java.ParserTests.bug1145(ParserTests.java:213)
    [junit] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [junit] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 
    [junit] 

BUILD FAILED
/home/sid/processing/build/build.xml:449: The following error occurred while executing this line:
/home/sid/processing/build/build.xml:537: The following error occurred while executing this line:
/home/sid/processing/java/build.xml:128: Test processing.mode.java.ParserTests failed

```

sampottinger commented:
Thanks @monkstone! I’ll take a look today.

sampottinger committed:
Update PDE_es.properties

sampottinger commented:
Hey @monkstone! I am still looking into it but I wasn't able to reproduce in Linux (Ubuntu 18.04.1) or Windows (10 via Paperspace). That said, I see that you are running `11.0.3` while AdoptOpenJDK is only on `11.0.2+9` and the OpenJDK PPA on `11.0.1+13`.

That in mind, I am wondering if the output format of `javac` may have changed? That's because both of your failed tests are on "expected compiler exceptions" and those tests are running a regex against stderr output. That seems mildly error prone ha ha! I'm also not sure that checking for the offending line here is useful considering that `JavaBuild` and `PreprocessingService` use the JDT. My guess is that those checks are just left over from pre-JDT times. 

So, I updated the tests to ensure that things fail to compile for those tests as expected. However, I removed the line check since I don't think it offers much value given how Processing later compiles files and interprets errors.

Can you give it another shot and let me know how it goes? Thanks!

sampottinger committed:
Fix brittle compilation failure tests in ParserTests.

In response to @monkstone feedback, removing the line checking for compilation failure in ParserTests. While ensuring that compilation fails when expected is useful, the line parsing is unhelpful since JavaBuild and PreprocessingService use the JDT. This only makes the test brittle due to regex over stderr.

sampottinger committed:
Merge branch 'antlr' of github.com:sampottinger/processing into antlr

sampottinger commented:
> That fixed it for me but for some reason I had to explicitly checkout the commit 99b4569 (weird), and also needed to install rsync on Archlinux. I look forward to doing some more experimenting.

That is very strange :( I'll go ahead create a fresh clone on a fresh linux vm and make sure there isn't anything else that crops up. Thanks so much @monkstone and sorry for the earlier inconvenience.

sampottinger committed:
Allow preprocessor to emit ImportStatements, fixing contrib libs.

Allow the preprocessor to emit ImportStatements, allowing the JDT to read extra libraries (including contributed ones) at compilation time.

sampottinger committed:
Merge pull request #30 from sampottinger/import_statement_antlr

Allow preprocessor to emit ImportStatements, fixing contrib libs.

sampottinger commented:
Note: Important merge from #30 just completed - fixes use of contributed libraries.

sampottinger committed:
Minor style cleanup and bug fix in issue simplifier.

Minor stylistic cleanup in issue message simplifier as well as a fix with line attribution for missing name / type.

sampottinger committed:
Some minor fixes in missing var name.

sampottinger committed:
Merge branch 'antlr' of github.com:sampottinger/processing into antlr

sampottinger committed:
Simplify finding nearby token for error message.

sampottinger committed:
Fix MessageSimplifierUtilTest

sampottinger committed:
Fix inconsistent line fix on certain missing characters.

sampottinger committed:
Minor refactor for cleanliness plus added missing generic test.

sampottinger committed:
Conformancy with existing test.

sampottinger committed:
Minor fix in line offsets.

sampottinger committed:
Minor fixes in line placement within JavaBuild given new ANTLR.

sampottinger committed:
Allow multiple issue emit from PdePreprocessor for grammar error.

sampottinger committed:
Minor docstring update.

sampottinger commented:
Hey there! Sorry for a little noise. Just made a few (mostly minor) fixes after some testing today. I think noise should reduce now.

monkstone commented:
I've been doing a bit of testing with the latest build, and for some reason calling `size()` on ArrayList/List gives syntax error in processing ide. Further when `void mousePressed()` is included in code the ide complains about reducing visibility of inherited PApplet method. I hacked Shiffman box2d as follows to get it to work (replaced for loop, and made mousePressed explicitly public).
```java
  // We are setting a custom gravity
  box2d.setGravity(0, -20);

  // Create ArrayLists	
  polygons = new ArrayList<>();
  boundaries = new ArrayList<>();

  // Add a bunch of fixed boundaries
  boundaries.add(new Boundary(width/4,height-5,width/2-50,10,0));
  boundaries.add(new Boundary(3*width/4,height-50,width/2-50,10,0));
  boundaries.add(new Boundary(width-5,height/2,10,height,0));
  boundaries.add(new Boundary(5,height/2,10,height,0));
}

void draw() {
  background(255);

  // We must always step through time!
  box2d.step();

  // Display all the boundaries
  for (Boundary wall: boundaries) {
    wall.display();
  }

  // Display all the people
  for (CustomShape cs: polygons) {
    cs.display();
  }

  // people that leave the screen, we delete them
  // (note they have to be deleted from both the box2d world and our list
  Iterator<CustomShape> shapeIterator = polygons.iterator();
  while (shapeIterator.hasNext()) {
    CustomShape cs = shapeIterator.next();
    if (cs.done()) {
      shapeIterator.remove();
    }
  }
}

void mousePressed() {
  CustomShape cs = new CustomShape(mouseX,mouseY);
  polygons.add(cs);
}
```

sampottinger commented:
Hey @monkstone! I’ll take a look today thanks for reporting!

sampottinger committed:
Fixed special methods processing in antlr traversal.

Updated processing of size to be consistent with current processing and added additional methods to the "special" methods list requiring visibility modification. In response to @monkstone. Thanks @monkstone!

sampottinger committed:
Merge pull request #32 from sampottinger/fix_antlr_speical_method

Fixed special methods processing in antlr traversal.

sampottinger commented:
Hey @monkstone! Thank you so much for sticking with me here and apologies for that slightly embarrassing oversight.

There was an issue with the way that "special" methods were processed and, in part, this was caused by incompatibility with the new-ish `settings` behavior on mainline processing. I have resolved these issues, added tests to ensure they are caught in case of regression in the future, and can confirm that I can run the unmodified example code from https://github.com/shiffman/Box2D-for-Processing/tree/master/Box2D-for-Processing/dist/box2d_processing/examples/Boxes with this updated branch.

<img width="1158" alt="Screen Shot 2019-04-04 at 11 10 07 AM" src="https://user-images.githubusercontent.com/110391/55579197-8c611300-56cc-11e9-93c1-a107240a5996.png">

sampottinger committed:
Minor fix in error reporting message for extraneous input.

sampottinger committed:
Minor fix in ExtraneousInputMessageSimplifierStrategy.

sampottinger committed:
Speed up preprocessing service.

sampottinger committed:
Allow users to do weird stuff with methods matching special names.

sampottinger committed:
Allow calling of strangly named special method matches.

sampottinger committed:
Fixed loose print

sampottinger committed:
Minor typo fix.

sampottinger committed:
Minor edits in error message text for clarification.

sampottinger committed:
Have default visibility modifier as public.

For consistency with prior processing, disallow use of the default visbility and instead prepend public modifier if no modifier is given. In response to @dzaima.

sampottinger committed:
Allow for inclusion of block comment in calls to size.

sampottinger committed:
Merge pull request #38 from sampottinger/antlr_access_and_static

Sync #37 into ANTLR

sampottinger committed:
Pass all imports to JDT-related work as ImportStatements.

In response to @dzaima, fix IDE import errors by passing all imports into the JDT even if they are listed in code as result of preprocessor.

sampottinger committed:
Fixes for default visiblity annotated functions + fullscreen.

In response to @dzaima feedback, added support for inserting public into method with non-visiblity modifiers (previously undefined behavior) and added support for full screen. Related to processing#5753 but part of sampottinger#15.

sampottinger committed:
Some fixes on fullscreen in global context.

sampottinger committed:
Clean up preprocessor comment injection.

sampottinger committed:
Merge pull request #40 from sampottinger/antlr_various_fixes

Antlr various fixes

sampottinger commented:
Humm... #40 might have had adverse effect. Seeing some `public public` action. Taking a look...

sampottinger committed:
Fix issues related to sampottinger#37.

There were some issues related to public modifier injection that were introduced by sampottinger#37, causing a large number of sketches to fail. This impacts sampottinger#15 and does not impact but is related to #5753. To prevent regressions, also introduces JDT based tests to ensure edits are valid.

sampottinger committed:
Fix insertion of public on first token.

sampottinger committed:
Merge pull request #42 from sampottinger/antlr_public_fix

[antlr branch] Fix issues related to sampottinger#37.

sampottinger commented:
Hey there! Note that some important stuff merged in recently. Please re-pull before reporting issues here. Thanks!

sampottinger committed:
Speed up preprocessing service.

sampottinger committed:
Revert speed up preprocessing service.

sampottinger committed:
Fix color transparency.

sampottinger committed:
Make SyntaxProblem a kind of JavaProblem.

sampottinger committed:
Remove loose print statement.

sampottinger committed:
Living dangerously. Moved to antlr 4.7

sampottinger committed:
Fix automated code folder imports.

sampottinger committed:
Fix type erasure over preprocessor write.

sampottinger committed:
Minor formatting fix on sampottinger#45.

sampottinger committed:
Minor formatting fix.

sampottinger committed:
Merge branch 'antlr' into fix_antlr_code_folder

sampottinger committed:
Merge pull request #46 from sampottinger/fix_antlr_code_folder

Fix automated code folder imports.

sampottinger committed:
Fix escaped characters for character literals.

sampottinger committed:
Merge pull request #49 from sampottinger/antlr_fix_char_to_antlr

 Fix escaped characters for character literals.

sampottinger committed:
Remove deprecated AntlrInputStream.

sampottinger committed:
Allow genercised token stream in preprocessing.

sampottinger committed:
Broken use of optimized Java grammar.

sampottinger committed:
Migrated to optimized Java 8 grammar.

sampottinger committed:
Check for non method declaration containing size call.

sampottinger committed:
Allow funky char literals.

sampottinger committed:
Merge branch 'antlr' into antlr_fix_performance_to_antlr

sampottinger commented:
Hey @dzaima! Really excited to share some changes with you. See below:

<br>

> Also the live editor still shows errors on classes from code/ jar files though it runs fine.

I believe this is resolved with #46.

> Another live error fail: char c = '\n';

I believe this is resolved with #49.

> and it takes ~20 seconds

Wow this was more than just a little bit of a bummer. I ended up needing to change the Java grammar definitions :(. It's a long story but involves this: https://stackoverflow.com/questions/32910013/antlr-v4-java8-grammar-outofmemoryexception. Anyway, the bad news is that it is possible there may be a regression or two not captured yet by unit tests. The good news is that is MUCH more performant.

<img width="1045" alt="Screen Shot 2019-04-08 at 10 56 01 PM" src="https://user-images.githubusercontent.com/110391/55777060-9def4580-5a53-11e9-9ed9-9e9c781f4762.png">

I'll continue testing but this is all going to get merged in with #50 / #51.

<br>

Thanks so much for your help @dzaima! Also, have things been going well @monkstone?

sampottinger committed:
Merge pull request #51 from sampottinger/antlr_fix_performance_to_antlr

[antlr] A number of performance improvements for ANTLR including new grammar

sampottinger commented:
Thanks @dzaima! I’ll look into that today. Reallllyy happy it is faster though thanks for confirming that.

sampottinger committed:
Minor cleanup for handeling automated import statements.

sampottinger committed:
Merge branch 'antlr' of github.com:sampottinger/processing into antlr

sampottinger committed:
Minor refactor in the PdePreprocessor for readability.

sampottinger commented:
Thanks again for all of your help @dzaima! Should be fixed with #52.

<img width="250" alt="Screen Shot 2019-04-09 at 10 08 59 AM" src="https://user-images.githubusercontent.com/110391/55823612-f1df4600-5ab6-11e9-9f95-aa998d01fee5.png">

sampottinger committed:
Remove loose print statement in Editor.java

sampottinger committed:
Updated some missing curly given new grammar.

sampottinger committed:
Removed extraneous print and fixed some error msg.

Removed extranous print and fixed some error message generation given regressions caused by sampottinger#50 and sampottinger#51.

sampottinger committed:
Merge pull request #55 from sampottinger/antlr_remove_loose_print

[antlr] Number of error message generation fixes

sampottinger commented:
Error spamming should be fixed! Thanks @dzaima.

dzaima commented:
Yep, with the latest fixes the editor seems pretty stable! Will continue to report problems here as I encounter them working on stuff.

dzaima commented:
`@Override` without `public` is broken again: `@Override void setup() { }`

sampottinger commented:
Hey @dzaima!

> @Override without public is broken again

Humm there's a unit test for that now but maybe there's an issue with the JDT edit. I'll take a look.

> using the import functionality from the error tab on Map<Long, Long> map;

I'll try it out

sampottinger committed:
Fix append public when annotation is present for JDT edit.

sampottinger committed:
Merge branch 'java11' into antlr

sampottinger commented:
Hey @dzaima! Thank you for that bug report. I have the fixes ready for you but am waiting for #56 to clear CI.

> `@Override` without public is broken again

This regression was caused by an issue with the JDT edit emitted. Should be resolved now.

> using the import functionality from the error tab on Map<Long, Long> map;

Crazy enough, I think this was caused by modules in Java 11? I have resolved this issue on the `java11` and `antlr` branches. It will be synced to master with #56.

Thanks!

sampottinger committed:
Minor change in error description.

sampottinger commented:
#56 is merged!

sampottinger committed:
Expose width and height of sketch on preprocessor result.

sampottinger committed:
Provide default width / height for preprocessor result.

sampottinger committed:
Fix typo on ExtraneousInputMessageSimplifierStrategy.

sampottinger committed:
Allow PdePreprocessor clients to specify code folder imports flexibly.

Allow PdePreprocessor clients to provide code folder imports with and without the wildcard appended, offering some abstraction to client code both within Processing itself and by outside users.

sampottinger committed:
Merge pull request #59 from sampottinger/antlr_fix_rename

[antlr] Allow PdePreprocessor clients to specify code folder imports flexibly.

sampottinger committed:
Merge branch 'antlr' into processing-master

sampottinger committed:
Revert to antlr java/build.xml

sampottinger committed:
Merge pull request #63 from sampottinger/processing-master

Update antlr branch with processing-master

anonymix007 commented:
If I try to print some unicode letters, π for example, then it just prints ? else, println('π'+0) gives 63.
In original Processing println('π'+0) gives 960 in console. println('π') gives ?.

JDK11:
![image](https://user-images.githubusercontent.com/48598263/57541080-e6519a00-7356-11e9-97c6-f0e93254766f.png)

Original:
![image](https://user-images.githubusercontent.com/48598263/57541716-5f9dbc80-7358-11e9-9ebc-c928d5a68c7a.png)

dzaima commented:
@anonymix007 For me, `println('π')` displays correctly in the console, `println('π'+0)` prints `960`, and `π` shown with `text()` draws correctly on Linux.

From the `63` it seems that `π` in the source is getting literally replaced with `?` - no idea why that'd be different between OSes - possibly wrong encoding usage? If you save a sketch with `π` in the source, does closing & reopening the sketch keep it? If you export the sketch, does `π` appear in `sources/projectName.java`, or is it already a `?`?

anonymix007 commented:
@dzaima Closing & reopening keeps π on place. After exporting sketch π replaced with ?, but in *.pde π symbol isn't replaced.

Original does well except for console:
![image](https://user-images.githubusercontent.com/48598263/57585910-f4d6b780-74f6-11e9-8bf1-dd97f6dd4fef.png)
And JDK11 one:
![image](https://user-images.githubusercontent.com/48598263/57586091-3d8f7000-74f9-11e9-989a-526338ba6aed.png)

dzaima commented:
https://github.com/sampottinger/processing/blob/e3c965bcb534d1160d6cff44780fdeef007ba9d8/java/src/processing/mode/java/JavaBuild.java#L232

This line should be changed to
```java
final PrintWriter stream = new PrintWriter(new FileWriter(java, StandardCharsets.UTF_8));
```
to fix the above unicode issue for Windows.

sampottinger commented:
Hello! Sorry for going incognito there. I’ve been trying to finish up a paper submission before the end of my sabbatical (I might still be a bit slow to respond). Thank you very much @dzaima and @anonymix007! I will make this change.

sampottinger committed:
Use UTF8 in writing preprocessor result

Fix handling of unicode characters after preprocessor.

Incorporates @dzaima feedback to address issue reported by @anonymix007 regarding the treatment of unicode characters through the preprocessor.

sampottinger committed:
Add missing import re #72

sampottinger committed:
Merge pull request #72 from sampottinger/antlr-utf8

Use UTF8 in writing preprocessor result

sampottinger commented:
Thank you very much @dzaima and @anonymix007! This is resolved.

dzaima commented:
What's with the `out.addEmptyLine`s in [this commit](https://github.com/sampottinger/processing/commit/675af86da59987ccbdb3eef33f1ff0f6c519d7a9)? (let me guess - didn't uncheck replacing everywhere. such an annoying feature sometimes)

More specifically, it's one of the things (besides what's already wrong in official master) that prevents tweak mode from working (i did manage to get it to work here after fixing things):

https://github.com/sampottinger/processing/blob/675af86da59987ccbdb3eef33f1ff0f6c519d7a9/java/src/processing/mode/java/tweak/TweakClient.java#L142

sampottinger commented:
Hey @dzaima! So sorry for the delay. Releasing fix presently...

sampottinger commented:
Oh my... it looks like the tweaks issues are present on mainline as well. I think I probably should tackle that separately from this PR / https://github.com/processing/processing/pull/5753. If I get time, I'll take a look on mainline.

dzaima commented:
@sampottinger tweak mode not working is already [reported](https://github.com/processing/processing/issues/5805) on mainline, I was just mentioning a separate, independent problem with the version here. Of course the mainline should also be fixed, but this should too.

dzaima committed:
Clean up comments

sampottinger committed:
Merge pull request #75 from dzaima/antlr

Clean up comments

sampottinger committed:
Merge pull request #81 from sampottinger/io_refactor_revert

Revert of sampottinger/processing #3 for the antlr branch.

jeremydouglass commented:
A new pull proposes a fix for tweak mode on master: #5909

sampottinger commented:
Hey @jeremydouglass thanks for the heads up! Also thank you @galsasson. :D I'm still deciding how best to tackle PRs unrelated to Java 11 / new ANTLR but, after I give it just a little thought, I might merge into my master without having it go into https://github.com/processing/processing/pull/5753. It would be nice to test tweaks mode with my branch.

sampottinger committed:
Minor stylistic cleanup in PdeIssueEmitter.java

Minor stylistic cleanup in PdeIssueEmitter.java, keeping identical base class method implementations in place and fixing long lines.

sampottinger committed:
Merge pull request #92 from sampottinger/antlr_clean_listener

Minor stylistic cleanup in PdeIssueEmitter.java

sampottinger committed:
Bump memory for Processing and default sketch memory.

We are seeing some sketches exceed the heap limits either through the JDT or ANTLR when they are very large (>50k lines). This PR increases the heap for helping the IDE keep up when the parse trees get larger. Separately, looking at benchmarks online, Java 11 is super fast but it may be a little less prudent on the heap. So, increasing the default sketch max memory to match the IDE.

sampottinger committed:
Fix linux memory increase for IDE.

sampottinger committed:
Merge pull request #94 from sampottinger/antlr_memory

Bump memory for Processing and default sketch memory.

sampottinger committed:
Minor fix in error line attribution for ANTLR.

sampottinger committed:
Merge pull request #97 from sampottinger/antlr_line_correction

Minor fix in error line attribution for ANTLR.

sampottinger committed:
Implemented check for imports in class bodies.

Implemented check for imports in class bodies without localization, providing a more informative error message when imports end up in unexpected places.

sampottinger committed:
Added localization for import statment inside class body.

sampottinger committed:
Add missing PdeParseTreeErrorListener.java

sampottinger committed:
Merge pull request #101 from sampottinger/antlr_detect_bad_import

Detect import within class body.

sampottinger committed:
Opened access to Java 11 language features.

Moved from Java 8 to Java 11 for Compiler and JDT. Also added gramatical support for local variable type inference (var).

sampottinger committed:
Merge pull request #105 from sampottinger/antlr_var

Opened access to Java 11 language features.

jeremydouglass commented:
re: 8e0bf4e in this PR, is the idea for this PR that PDE would initially migrate to Java 8 language features, or that it would skip straight to Java 11 language features?

sampottinger commented:
Originally we were anticipating going to Java 8 and then reaching for Java 11 later. Based on some community feedback, broader documentation within the Java community has started reflecting language features in Java 11 so it may be confusing for folks to say try using lambdas or for-colon loops but need to understand that `var` was introduced in Java 9. I think saving folks from that complexity is a goal inline with #15. The only downside that has emerged so far: this widens the gap for android processing which, as of writing, doesn't support Java 11 grammar. At the same time, [Java 8 features aren't totally safe in the Android community right now anyway](https://developer.android.com/studio/write/java8-support).

That in mind, we can revert #105 if there is a compelling reason. Let me know what you think!

sampottinger commented:
^ Ah dang sorry forgot to at-mention you @jeremydouglass in the above.

jeremydouglass commented:
Very excited about Java 11, but I don't actually have a strong opinion -- if the goal is to eventually merge into upstream rather than stay forked, I'm guessing that is something to strategize with e.g. benfry, as either decision will have lots of implications for documentation and support etc. Given the installed user base, I could see arguments for either a gradual approach or an all-at-once transition. Best decision is probably whichever one sounds most survivable to the core devs....

sampottinger commented:
Thanks @jeremydouglass! I completely agree and thank you again for your continued support through this. I've really been struggling with what kind of decisions this fork should feel are in purview (something similar also surfaced for #103 which relates to https://github.com/processing/processing/issues/3199 where I took the more conservative approach)... So, I am open to suggestions on this. My current thinking:

 - On one hand https://github.com/processing/processing/issues/3054 is about 4 years old and predates Java 11 so one feasible interpretation of its intent is to deliver an up to date Java grammar for Processing and that it would have said Java 11 had it been released at the time. I also empathize with the sentiment from #104. Online documentation and expectations for what it means to develop with Java have also shifted so the goal posts for delivering a positive experience for future Java developers is also changing.
 - On the other side, even if time between communications can be quite long, I would strongly prefer to be making these decisions with deeper connection to Processing's leadership. 

After giving this some more thought, I probably wont revert #105 for now but will certainly be open to reversing if there's strong opinion in the opposite direction from the core devs / Ben. Ultimately, I don't see it as introducing too much risk for now (especially since we can't run on Java 8 anymore anyway so we'd only be limiting the API availability but couldn't buy any flexibility in targets).

jeremydouglass commented:
Absolutely! Sounds good, and no question you have a strong strategy and the logic makes sense to me. Just suggesting that the 8 v 11 decision might be a key point -- if not the key point -- for coordination and communication if you have merging in mind.

sampottinger commented:
Totally! When talking with @benfry, will definitely walk through that.

sampottinger commented:
(Sorry ... less for you @jeremydouglass and more for googlers from the future...) Just to clarify for progeny, it was decided in https://github.com/processing/processing/pull/5753 that we would be moving to Java 11 in terms of the VM. The only question is if the _grammar_ should support Java 11 or Java 8 features within the IDE. The Processing code base within this fork will not compile under the Java 8 JDK due to backwards incompatible changes made within the Java APIs. It is only intended to work with OpenJFX and OpenJDK 11+... There have been no efforts made to ensure the code base has a path to build under either OpenJDK 8 or Oracle JDK 8. See https://github.com/processing/processing/pull/5753 for further details.

sampottinger committed:
Fixed license header in .g4 files.

sampottinger committed:
Merge pull request #109 from sampottinger/antlr_update_grammar_license

Fixed license header in .g4 files.

sampottinger commented:
Hey there! So... I spoke with Ben! We are still working through some pieces for getting this into mainline but, on this question specifically, we are going to hold on to the Java 11 language compatability changes for now.
